benchmarks/distributed/rpc/ps/benchmark.py

```diff
@@ -0,0 +1,384 @@
+from functools import wraps
+import os
+import random
+import time
+import threading
+import copy
+import sys
+
+import torch
+import torch.distributed as dist
+import torch.distributed.autograd as dist_autograd
+from torch.distributed.optim import DistributedOptimizer
+import torch.distributed.rpc as rpc
+from torch.distributed.rpc import RRef
+from torch.distributed.rpc import TensorPipeRpcBackendOptions
+import torch.multiprocessing as mp
+from torch.nn.parallel import DistributedDataParallel as DDP
+import torch.optim as optim
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.distributed as c10d
+import torch.distributed.rpc as rpc
+
+GLOO = "gloo"
+NCCL = "nccl"
+
+# --------------------------- privates -----------------------------------------
+
+
+def _get_name(rank, world_size):
+    if rank < world_size - 2:
+        return "trainer{}".format(rank)
+    elif rank == world_size - 2:
+        return "gs"
+    else:
+        return "master"
+
+
+def _call_method(method, rref, *args, **kwargs):
+    return method(rref.local_value(), *args, **kwargs)
+
+
+def _remote_method(method, rref, *args, **kwargs):
+    args = [method, rref] + list(args)
+    return rpc.rpc_async(rref.owner(), _call_method, args=args, kwargs=kwargs)
+
+# --------------------------- Parameter Server -----------------------------------------
+
+
+class GradientServer(nn.Module):
+
+    def __init__(self, world_size):
+        super().__init__()
+        torch.manual_seed(0)
+        self.lock = threading.Lock()
+        self.rank_gradient_counters = [[0] * world_size, [0] * world_size]
+        self.gradient_dicts = [{}, {}]
+        self.gradient_dim = {}
+
+    # batch processing just store the gradients
+    # trainer and server update every iteration
+    # batch case - 1 cuda kernel
+    # iteration case - n cuda kernels
+
+    @rpc.functions.async_execution
```

Comment:
functions decorated with `@rpc.functions.async_execution` needs to return a `Future` object.